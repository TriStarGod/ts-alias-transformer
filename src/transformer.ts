import ts from 'typescript';
import _ from 'lodash';

export function transformer<T extends ts.Node>(
  program: ts.Program
): ts.TransformerFactory<T> {
  const checker = program.getTypeChecker();

  return context => {
    const visit: ts.Visitor = node => {
      node = ts.visitEachChild(node, visit, context);
      /*
        Convert type references to type literals
  
          interface IUser {
            username: string
          }
          type User = IUser <--- IUser is a type reference
          interface Context {
            user: User <--- User is a type reference
          }
  
        In both cases we want to convert the type reference to
        it's primitive literals. We want:
  
          interface IUser {
            username: string
          }
          type User = {
            username: string
          }
          interface Context {
            user: {
              username: string
            }
          }
      */
      if (ts.isTypeReferenceNode(node)) {
        const symbol = checker.getSymbolAtLocation(node.typeName);
        const type = checker.getDeclaredTypeOfSymbol(symbol);
        const declarations = _.flatMap(checker.getPropertiesOfType(type), property => {
          /*
            Type references declarations may themselves have type references, so we need
            to resolve those literals as well 
          */
          return _.map(property.declarations, visit);
        });
        return ts.createTypeLiteralNode(declarations.filter(ts.isTypeElement));
      }

      /* 
        Convert type alias to interface declaration
          interface IUser {
            username: string
          }
          type User = IUser
      
        We want to remove all type aliases
          interface IUser {
            username: string
          }
  
          interface User {
            username: string  <-- Also need to resolve IUser
          }
      
      */

      if (ts.isTypeAliasDeclaration(node)) {
        const symbol = checker.getSymbolAtLocation(node.name);
        const type = checker.getDeclaredTypeOfSymbol(symbol);
        const declarations = _.flatMap(checker.getPropertiesOfType(type), property => {
          // Resolve type alias to it's literals
          return _.map(property.declarations, visit);
        });

        // Create interface with fully resolved types
        return ts.createInterfaceDeclaration(
          [],
          [ts.createToken(ts.SyntaxKind.ExportKeyword)],
          node.name.getText(),
          [],
          [],
          declarations.filter(ts.isTypeElement)
        );
      }
      // Remove all export declarations
      if (ts.isImportDeclaration(node)) {
        return null;
      }

      return node;
    };

    return node => ts.visitNode(node, visit);
  };
}
