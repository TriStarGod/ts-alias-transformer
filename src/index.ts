import path from 'path';
import ts from 'typescript';
import _ from 'lodash';
import fs from 'fs';
import { transformer } from './transformer';

const filePath = path.resolve(_.first(process.argv.slice(2)));

const program = ts.createProgram([filePath], {});
const source = program.getSourceFile(filePath);
const printer = ts.createPrinter();

// Run source file through our transformer
const result = ts.transform(source, [transformer(program)]);

// Create our output folder
const outputDir = path.resolve(__dirname, '../generated');
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir);
}

// Write pretty printed transformed typescript to output directory
fs.writeFileSync(
  path.resolve(__dirname, '../generated/models.ts'),
  printer.printFile(_.first(result.transformed))
);
