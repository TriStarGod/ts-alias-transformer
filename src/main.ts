import path from 'path';
import ts from 'typescript';
import _ from 'lodash';
import fs from 'fs';
import { transformer } from './transformer';

export function main(modelsFilePath: string, outDirFilePath: string) {
  const program = ts.createProgram([modelsFilePath], {});
  const source = program.getSourceFile(modelsFilePath);
  const printer = ts.createPrinter();

  // Run source file through our transformer
  const result = ts.transform(source, [transformer(program)]);

  // Create our output folder
  if (!fs.existsSync(outDirFilePath)) {
    fs.mkdirSync(outDirFilePath);
  }

  // Write pretty printed transformed typescript to output directory
  fs.writeFileSync(
    path.resolve(outDirFilePath, './models.ts'),
    printer.printFile(_.first(result.transformed))
  );
}
