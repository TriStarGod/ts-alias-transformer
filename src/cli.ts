#!/usr/bin/env node
import program from 'commander';

import { main } from './main';

program
  .version('0.0.1')
  .description('resolve type aliases to their lowest structure')
  .option('-m, --models [path]', 'the path to your models')
  .option('-o, --outDir [path]', 'desired output directory');

program.parse(process.argv);

main(program.models, program.outDir);
