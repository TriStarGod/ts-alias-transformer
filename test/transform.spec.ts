import ts from 'typescript';
import path from 'path';
import { transformer } from '../src/transformer';

describe('ts-alias-transformer', () => {
  it('At least works with my example', () => {
    const filePath = path.resolve(__dirname, './models', 'basic_example.ts');
    const program = ts.createProgram([filePath], {});
    const source = program.getSourceFile(filePath);
    const printer = ts.createPrinter();

    const result = ts.transform(source, [transformer(program)]);
    expect(printer.printFile(result.transformed[0])).toMatchSnapshot();
  });
  it('works with a relatively complicated node_module type import', () => {
    const filePath = path.resolve(__dirname, './models', 'from_package.ts');
    const program = ts.createProgram([filePath], {});
    const source = program.getSourceFile(filePath);
    const printer = ts.createPrinter();

    const result = ts.transform(source, [transformer(program)]);
    expect(printer.printFile(result.transformed[0])).toMatchSnapshot();
  });
});
