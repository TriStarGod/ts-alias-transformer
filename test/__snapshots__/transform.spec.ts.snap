// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ts-alias-transformer At least works with my example 1`] = `
"export interface User {
    username: string;
    info: {
        firstName: string;
        lastName: string;
    };
}
export interface Todo {
    createdBy: {
        username: string;
        info: {
            firstName: string;
            lastName: string;
        };
    };
    text: string;
}
"
`;

exports[`ts-alias-transformer works with a relatively complicated node_module type import 1`] = `
"export interface Heap {
    total_heap_size: number;
    total_heap_size_executable: number;
    total_physical_size: number;
    total_available_size: number;
    used_heap_size: number;
    heap_size_limit: number;
    malloced_memory: number;
    peak_malloced_memory: number;
    does_zap_garbage: {
        toString(radix?: number): string;
        toFixed(fractionDigits?: number): string;
        toExponential(fractionDigits?: number): string;
        toPrecision(precision?: number): string;
        valueOf(): number;
        toLocaleString(locales?: string | string[], options?: {
            localeMatcher?: string;
            style?: string;
            currency?: string;
            currencyDisplay?: string;
            useGrouping?: boolean;
            minimumIntegerDigits?: number;
            minimumFractionDigits?: number;
            maximumFractionDigits?: number;
            minimumSignificantDigits?: number;
            maximumSignificantDigits?: number;
        }): string;
    };
}
"
`;

exports[`ts-alias-transformer works with type nesting 1`] = `
"export interface User {
    user: {
        username: string;
        info: {
            firstName: string;
            lastName: string;
        };
    };
    todo: {
        createdBy: {
            username: string;
            info: {
                firstName: string;
                lastName: string;
            };
        };
        text: string;
    };
}
"
`;
